src/node.cc

```diff
@@ -2487,26 +2491,48 @@ struct node_module* get_linked_module(const char* name) {
   return mp;
 }
 
-// DLOpen is process.dlopen(module, filename).
+// DLOpen is process.dlopen(module, filename, flags).
 // Used to load 'module.node' dynamically shared objects.
 //
 // FIXME(bnoordhuis) Not multi-context ready. TBD how to resolve the conflict
 // when two contexts try to load the same shared object. Maybe have a shadow
 // cache that's a plain C list or hash table that's shared across contexts?
 static void DLOpen(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
+  bool is_dlopen_error = false;
+  const char* dlerrmsg = nullptr;
   uv_lib_t lib;
 
   CHECK_EQ(modpending, nullptr);
 
-  if (args.Length() != 2) {
-    env->ThrowError("process.dlopen takes exactly 2 arguments.");
+  if (args.Length() < 2) {
+    env->ThrowError("process.dlopen needs at least 2 arguments.");
     return;
   }
 
+  int32_t flags = RTLD_LAZY;
+  if (args.Length() > 2 && !args[2]->Int32Value(env->context()).To(&flags)) {
+    return env->ThrowTypeError("flag argument must be an integer.");
+  }
+
   Local<Object> module = args[0]->ToObject(env->isolate());  // Cast
   node::Utf8Value filename(env->isolate(), args[1]);  // Cast
-  const bool is_dlopen_error = uv_dlopen(*filename, &lib);
+
+#ifdef __POSIX__
+  // libuv does not support passing flags to dlopen,
+  // so we need to open code the uv_dlopen implementation.
+  lib.errmsg = nullptr;
+  lib.handle = dlopen(*filename, flags);
+  if (!lib.handle) {
+    is_dlopen_error = true;
+    dlerrmsg = dlerror();
+  }
```

Comment:
dlclose-ing a shared object is safe only when you know it never executed any code (but we don't know that, there may be static constructors) or when you are absolutely sure it cleaned up after itself (which we don't know either.)